paperqa/clients/openalex.py

```diff
@@ -0,0 +1,238 @@
+
+        return await parse_openalex_to_doc_details(results_data)
+
+
+async def parse_openalex_to_doc_details(message: dict[str, Any]) -> DocDetails:
+    """Parse OpenAlex API response to DocDetails.
+
+    Args:
+        message: The OpenAlex API response message.
+
+    Returns:
+        Parsed document details.
+    """
+
+    # author_name will be FamilyName, GivenName Middle initial. (if available)
+    # there is no standalone "FamilyName" or "GivenName" fields
+    # this manually constructs the name into the format the other clients use
+    def reformat_name(name: str) -> str:
+        # https://regex101.com/r/74vR57/1
+        pattern = r"^([^,]+),\s*(.+?)(?:\s+(\w+\.?))?$"
+        match = re.match(pattern, name)
+        if match:
+            family_name, given_name, middle = match.groups()
+
+            family_name = family_name.strip()
+            given_name = given_name.strip()
+
+            reformatted = f"{given_name}"
+            if middle:
+                reformatted += f" {middle.strip()}"
+            reformatted += f" {family_name}"
+            return reformatted.strip()
+
+        return name
+
+    authors = [
+        authorship.get("raw_author_name")
+        for authorship in message.get("authorships", [])
+    ]
+    authors = [reformat_name(author) for author in authors]
+    sanitized_authors = [convert_acutes(author) for author in authors]
+
+    publisher = (
+        message.get("primary_location", {})
+        .get("source", {})
+        .get("host_organization_name")
+    )
+    journal = message.get("primary_location", {}).get("source", {}).get("display_name")
+
+    return DocDetails(  # type: ignore[call-arg]
+        key=None,
+        bibtex_type=BIBTEX_MAPPING.get(message.get("type", "other"), "misc"),
+        bibtex=None,
+        authors=sanitized_authors,
+        publication_date=datetime.fromisoformat(message.get("publication_date", "")),
+        year=message.get("publication_year"),
+        volume=message.get("biblio", {}).get("volume"),
+        issue=message.get("biblio", {}).get("issue"),
+        publisher=publisher,
+        issn=message.get("primary_location", {}).get("source", {}).get("issn_l"),
+        pages=message.get("biblio", {}).get("last_page"),
+        journal=journal,
+        url=message.get("doi"),
+        title=message.get("title"),
+        citation_count=message.get("cited_by_count"),
+        doi=message.get("doi"),
+        other=message,
+    )
+
+
+class OpenAlexProvider(DOIOrTitleBasedProvider):
+    async def get_doc_details(
+        self, doi: str, session: aiohttp.ClientSession
+    ) -> DocDetails | None:
+        """Get document details by DOI.
+
+        Args:
+            doi: The DOI of the document.
+            session: The active session of the request.
+
+        Returns:
+            The document details if found, otherwise None.
+        """
+        return await get_doc_details_from_openalex(doi=doi, session=session)
```

Comment:
Do we need `get_doc_details_from_openalex`? Can we just put the implementation here?

Reply:
I thought it was cleaner this way. Can change it though

Reply:
Oh okay if you like it then it's fine